### PR TITLE
[test] [amdgpu] Fix bug with allocs bb in function body

### PR DESCRIPTION
**Pull request by GaleSeLee**
Issue: # related https://github.com/taichi-dev/taichi/pull/7293 FIXME

### Brief Summary
For the sake of code simplicity, the pass is placed at the beginning of the jit instead of where `allocs` was created.

### CODE REVIEW

File: taichi/runtime/llvm/llvm_context_pass.h [hunk 1]
```diff
@@ -174,6 +174,39 @@ struct AMDGPUAddStructForFuncPass : public ModulePass {
...
+        if (!now_alloca ||
+            now_alloca->getType()->getAddressSpace() != (unsigned)0) {
+          continue;
+        }
+        alloca_inst_vec.push_back(now_alloca);
+      }
+      for (auto &allocainst : alloca_inst_vec) {
+        auto alloca_type = allocainst->getAllocatedType();
+        llvm::IRBuilder<> builder(allocainst);
+        auto *new_alloca = builder.CreateAlloca(alloca_type, (unsigned)5);
```
Comment (jim19930609) [inline taichi/runtime/llvm/llvm_context_pass.h:198]: specify the meaning of magic number `5` - what memory space does it represent?

File: taichi/runtime/llvm/llvm_context_pass.h [hunk 2]
```diff
@@ -174,6 +174,39 @@ struct AMDGPUAddStructForFuncPass : public ModulePass {
...
+        auto new_type = llvm::PointerType::get(alloca_type, (unsigned)0);
```
Comment (jim19930609) [inline taichi/runtime/llvm/llvm_context_pass.h:199]: ditto

File: taichi/runtime/llvm/llvm_context_pass.h [hunk 3]
```diff
@@ -174,6 +174,39 @@ struct AMDGPUAddStructForFuncPass : public ModulePass {
...
+  bool runOnFunction(llvm::Function &f) override {
+    for (auto &bb : f) {
+      if (bb.getName() != "allocs")
+        continue;
+
+      std::vector<AllocaInst *> alloca_inst_vec;
+      for (Instruction &inst : bb) {
+        AllocaInst *now_alloca = dyn_cast<AllocaInst>(&inst);
+        if (!now_alloca ||
+            now_alloca->getType()->getAddressSpace() != (unsigned)0) {
```
Comment (jim19930609) [inline taichi/runtime/llvm/llvm_context_pass.h:190]: ditto

File: taichi/runtime/llvm/llvm_context_pass.h [hunk 4]
```diff
@@ -174,6 +174,39 @@ struct AMDGPUAddStructForFuncPass : public ModulePass {
   }
 };
 
+struct AMDGPUConvertFunctionBodyAllocsAddressSpacePass : public FunctionPass {
```
Comment (jim19930609) [inline taichi/runtime/llvm/llvm_context_pass.h:177]: Add some comments to explain what modifications were applied to the original `AllocaInst`

**Review by jim19930609 — APPROVED** [review]

File: taichi/runtime/llvm/llvm_context_pass.h (hunk 4 above)
Comment (GaleSeLee) [reply taichi/runtime/llvm/llvm_context_pass.h:177]: Thanks for the discussion with @jim19930609, a better way is to use `hipfreeasync`. Due to the rocm version problem, I think we can switch to the `hipfreeasync` implementation afterwards.

**Pull request merged**